### ISSUE TITLE
False-positive `InvalidTemplateParam` with `@template-covariant T` when a callable parameter accepts `T`

### ISSUE THREAD

**Issue by Shira-3749** [vimeo/psalm#10848]
When a class defines `@template-covariant T` and one of the methods has a `callable` parameter which accepts `T` as an argument, I'm getting an `InvalidTemplateParam` error:

https://psalm.dev/r/2eb66f01fe

I'm not sure the error is valid in this case. (And it seems to [work fine](https://psalm.dev/r/3ebad94817) otherwise.)

PHPStan doesn't seem to mind this: https://phpstan.org/r/ac9be5aa-4437-40c3-b55d-3ffaa236cb67
But will also error when it's an actual issue: https://phpstan.org/r/96d33c99-1ff7-4541-b024-c8cfaea34e30

I've also tried it in C# (as a language with native generics support) and it seems to allow this as well:

```csharp
public interface IExample<out T>
{
    public void call(Action<T> callback);
}
```
(`out` is equivalent to `@template-covariant` here)

**Comment by weirdan** [comment]
Note to future me: `callable(T):void` should be valid in this context, but `callable():T` shouldn't.

**Comment by vkurdin** [comment]
It would be awesome if Psalm supported this Scala machinery on callables and variance: https://scastie.scala-lang.org/ZbzouFbvRHGPSwwB8OrNoA
`+A` stands for `@template-covariant A`
`-A` is `@template-contravariant A`
`Unit` is `void`
`A => Boolean` is `callable(A): bool`
`(A => Boolean) => Boolean` is `callable(callable(A): bool): bool`
